File: src/main/java/cz/vhromada/catalog/gui/commons/AbstractListDataModel.java
package cz.vhromada.catalog.gui.commons;

import java.util.List;

import javax.swing.AbstractListModel;

import cz.vhromada.result.Result;
import cz.vhromada.result.Status;

/**
 * An abstract class represents data model for list with data.
 *
 * @param <T> type of data
 * @author Vladimir Hromada
 */
public abstract class AbstractListDataModel<T> extends AbstractListModel<String> {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * List of data
     */
    private List<T> data;

    @Override
    public int getSize() {
        return data.size();
    }

    @Override
    public String getElementAt(final int index) {
        return getDisplayValue(getObjectAt(index));
    }

    /**
     * Returns data object at the specified index.
     *
     * @param index the requested index
     * @return data object at index
     */
    public T getObjectAt(final int index) {
        return data.get(index);
    }

    /**
     * Updates model.
     */
    public final void update() {
        final Result<List<T>> result = getData();

        if (Status.OK == result.getStatus()) {
            data = result.getData();
        } else {
            throw new IllegalArgumentException("Can't get data. " + result);
        }
    }

    /**
     * Returns result with data.
     *
     * @return result with data.
     */
    protected abstract Result<List<T>> getData();

    /**
     * Returns display value for data.
     *
     * @param dataObject data
     * @return display value for data
     */
    protected abstract String getDisplayValue(T dataObject);

}

File: src/main/kotlin/cz/vhromada/catalog/gui/common/AbstractListDataModel.kt
package cz.vhromada.catalog.gui.common

import cz.vhromada.validation.result.Result
import cz.vhromada.validation.result.Status
import javax.swing.AbstractListModel

/**
 * An abstract class represents data model for list with data.
 *
 * @param <T> type of data
 * @author Vladimir Hromada
 */
abstract class AbstractListDataModel<T> : AbstractListModel<String>() {

    /**
     * List of data
     */
    private lateinit var data: List<T>

    override fun getSize(): Int {
        return data.size
    }

    override fun getElementAt(index: Int): String {
        return getDisplayValue(getObjectAt(index))
    }

    /**
     * Returns data object at the specified index.
     *
     * @param index the requested index
     * @return data object at index
     */
    fun getObjectAt(index: Int): T {
        return data[index]
    }

    /**
     * Updates model.
     */
    fun update() {
        val result = getData()

        if (Status.OK == result.status) {
            data = result.data!!
        } else {
            throw IllegalArgumentException("Can't get data. $result")
        }
    }

    /**
     * Returns result with data.
     *
     * @return result with data.
     */
    protected abstract fun getData(): Result<List<T>>

    /**
     * Returns display value for data.
     *
     * @param dataObject data
     * @return display value for data
     */
    protected abstract fun getDisplayValue(dataObject: T): String

}
